### PR TITLE
Create-detrend-object

### PR DESCRIPTION
**Pull request by jExbrayat**
Create detrend.py file defining the detrend class. 
Create a temporary demo notebook for easier visualisation of the detrend class' aim.

### CODE REVIEW

File: src/detrend.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            x, y, s=smoothing_factor, k=degree
+        )  # s: smoothing factor
+
+        # Interpolate the prices
+        spline = interpolate.BSpline(t, c, k, extrapolate=False)
+        y_interpolate = spline(x)
+
+        return y_interpolate, parameters_output
+
+    def __init__(
```
Comment (Ab2nour) [inline src/detrend.py]: init should be the first method of the class

File: src/detrend.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    def predict(self) -> np.ndarray:
+        """_summary_
+
+        Returns:
+            np.ndarray: detrended values, 1 dimensional array of length len(y)
+        """
+        self.y_predict = self.y_original - self.fitted_values
+        return self.y_predict
+
+    def fancy_plot(self, xticklabels=None) -> None:
```
Comment (Ab2nour) [inline src/detrend.py]: add type hint for xticklabels

**Review by Ab2nour — CHANGES_REQUESTED** [review]

**Comment by jExbrayat** [comment]
>  ok boomer

**Comment by jExbrayat** [comment]
> init should be the first method of the class 

sorry boomer

File: src/detrend.py (hunk 2 above)
Comment (Ab2nour) [reply src/detrend.py]: ok :)

File: src/detrend.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
+from typing import Callable
+from sklearn.linear_model import LinearRegression
+from statsmodels.tsa.deterministic import DeterministicProcess
+from scipy import interpolate
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+
+
+class detrend:
```
Comment (Ab2nour) [inline src/detrend.py]: class names should be PascalCase

**Review by Ab2nour — CHANGES_REQUESTED** [review]

**Review by Ab2nour — APPROVED** [review]

File: src/detrend.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+class Detrend:
+    def __init__(
+        self,
+        method: str = "LinearRegression",
+        poly_order: int = 2,
+        n_segments: int = 5,
+        window: int = 100,
+        alpha: float = 0.05,
+        bsplines_factors: tuple = (10, 3),
```
Comment (Ab2nour) [inline src/detrend.py]: type hint: set the size of the tuple
for example, tuple of 2 ints should be type hinted this way:
```py
bsplines_factors: tuple[int, int] = (10, 3),
```

Comment (Ab2nour) [inline src/detrend.py]: type hint: set the size of the tuple
for example, tuple of 2 ints should be type hinted this way:
```py
bsplines_factors: tuple[int, int] = (10, 3),
```

Comment (Ab2nour) [inline src/detrend.py]: type hint: set the size of the tuple
for example, tuple of 2 ints should be type hinted this way:
```py
bsplines_factors: tuple[int, int] = (10, 3),
```

Comment (Ab2nour) [inline src/detrend.py]: type hint: set the size of the tuple
for example, tuple of 2 ints should be type hinted this way:
```py
bsplines_factors: tuple[int, int] = (10, 3),
```

File: src/detrend.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    def predict(self) -> np.ndarray:
+        """_summary_
+
+        Returns:
+            np.ndarray: detrended values, 1 dimensional array of length len(y)
+        """
+        self.y_predict = self.y_original - self.fitted_values
+        return self.y_predict
+
+    def fancy_plot(self, xticklabels: pd.core.indexes.base.Index | None = None) -> None:
```
Comment (Ab2nour) [inline src/detrend.py]: add docstring

File: src/detrend.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        """
+        smoothing_factor = self.smoothing_factor
+        degree = self.degree
+        parameters_output = {
+            "Smoothing factor": smoothing_factor,
+            "Degree": degree,
+        }
+
+        # Define x and y
+        x = np.arange(len(y))  # time dummy, x
```
Comment (Ab2nour) [inline src/detrend.py]: maybe x and y should instead be named time_dummy and price?

**Review by Ab2nour — CHANGES_REQUESTED** [review]

File: src/detrend.py (hunk 4 above)
Comment (jExbrayat) [reply src/detrend.py]: done

File: src/detrend.py (hunk 6 above)
Comment (jExbrayat) [reply src/detrend.py]: done

File: src/detrend.py (hunk 5 above)
Comment (jExbrayat) [reply src/detrend.py]: done

**Comment by Ab2nour** [comment]
I merged [refactor/separate-methods-in-classes](https://github.com/Ab2nour/stock-analysis/pull/13/commits/b1bf723bc6c4f434ac4cc2e397243bb674538299) into this branch.

This branch also looks good to me, you can merge it into main, if you are finished @jExbrayat .

**Pull request merged**